apps/api/src/domain/models/shared/access-control/access-control-list.entity.ts

```diff
@@ -0,0 +1,61 @@
+import { InternalError } from '../../../../lib/errors/InternalError';
+import { DTO } from '../../../../types/DTO';
+import { AggregateId } from '../../../types/AggregateId';
+import { isNullOrUndefined } from '../../../utilities/validation/is-null-or-undefined';
+import BaseDomainModel from '../../BaseDomainModel';
+
+export class AccessControlList extends BaseDomainModel {
+    readonly allowedUserIds: AggregateId[];
+
+    readonly allowedGroupIds: AggregateId[];
+
+    /**
+     * TODO Consider a migration to make this property optional on all
+     * resources.
+     */
+    constructor(dto?: Partial<DTO<AccessControlList>>) {
+        super();
+
+        if (isNullOrUndefined(dto)) {
+            this.allowedUserIds = [];
+
+            this.allowedGroupIds = [];
+
+            return;
+        }
+
+        const { allowedUserIds: allowedUsers, allowedGroupIds: allowedGroups } = dto;
+
+        this.allowedUserIds = isNullOrUndefined(allowedUsers) ? [] : [...allowedUsers];
+
+        this.allowedGroupIds = isNullOrUndefined(allowedGroups) ? [] : [...allowedGroups];
+    }
+
+    allowUser(userId: AggregateId): AccessControlList {
+        if (this.allowedUserIds.includes(userId)) {
```

Comment:
Do we really want to throw? If so- make a note about it being the caller's responsibility to check this.